test/zdtm/static/cgroup_yard.hook

```diff
@@ -29,42 +30,22 @@ if sys.argv[1] == "--pre-dump":
 			os.mkdir(yard + "/" + ctrl)
 			subprocess.check_call(["mount", "-t", "cgroup", "none", yard + "/" + ctrl, "-o", opts])
 
-if sys.argv[1] == "--post-restore":
-	'''
-	Clean up the cgroup yard created during `--pre-dump`
-	'''
-	with open("/proc/self/cgroup") as f:
-		for line in f:
-			cgr = line.split(":")[1]
-
-			if cgr == "":
-				continue
-
-			if cgr.startswith("name="):
-				ctrl = cgr[len("name="):]
-			else:
-				ctrl = cgr
-
-			subprocess.check_call(["umount", yard + "/" + ctrl])
-			os.rmdir(yard + "/" + ctrl)
-	os.rmdir(yard)
-
 if sys.argv[1] in ["--pre-restore", "--clean"]:
 	'''
 	Clean up the leftover cgroups created by the test
 	'''
 	tname = tempfile.mkdtemp()
 	subprocess.call(["mount", "-t", "cgroup", "none", tname, "-o", "none,name=zdtmtst"])
 
-	try:
-		os.rmdir(os.path.join(tname, "subcg00", "subsubcg"))
-	except:
-		pass
-
-	try:
-		os.rmdir(os.path.join(tname, "subcg00"))
-	except:
-		pass
+	for cg in [os.path.join(tname, "subcg00", "subsubcg"),
+			os.path.join(tname, "subcg00")]:
+		if os.access(cg, os.F_OK):
+			os.rmdir(cg)
 
 	subprocess.call(["umount", tname])
 	os.rmdir(tname)
+
+if sys.argv[1] == "--clean":
+	if os.access(yard, os.F_OK):
+		subprocess.call(["umount", "-l", yard])
```

Comment:
If there are still some fds open, lazy umount will success and we will not notice the hanging fds. Wouldn't `umount -R` be a better idea? It should fail if there are some fds open.

Reply:
clean has to be always successful. All errors should be detected in other hooks (post-restore, post-dump).

Reply:
Actually, we can leave your code which umounts cgroups in post-restore. Is it enough to address a case what you are thinking about?

And we check file descriptors in zdtm.py, you can look at get_visible_state() for more details.

Reply:
I don't want to use umount -R, because it reads /proc/mounts, but there is a known race in the linux kernel when /proc/mounts doesn't show some mounts or show some mounts twice.

Reply:
If we check fds in zdtm.py, then I think all options are good. `umount -l` is the cleanest one I think.